driver-core/src/main/com/mongodb/operation/CommandOperationHelper.java

```diff
@@ -222,6 +222,10 @@ public T call(final ConnectionSource source) {
         } catch (MongoException e) {
             exception = e;
 
+            if (e instanceof MongoSocketException) {
+                binding.getSessionContext().markSessionDirty();
```

Comment:
This seems like a bad interaction with the NoOpSessionContext, which will throw UnsupportedOperationException.  Is there any guarantee that we don't have a NoOpSessionContext when we arrive here?  Seems safer for NoOpSessionContext to make this call a no-op rather than throw. Either that or protect all calls with `hasSession()`